returnn/forward.py

```diff
@@ -0,0 +1,213 @@
+        self._device = device
+
+        self.out_returnn_config_file = self.output_path("returnn.config")
+
+        self.out_hdf_files = {}
+        hdf_outputs = hdf_outputs if hdf_outputs else []
+        for output in hdf_outputs:
+            self.out_hdf_files[output] = self.output_path(output)
+        if not eval_mode:
+            self.out_hdf_files["output.hdf"] = self.output_path("output.hdf")
+            self.out_default_hdf = self.out_hdf_files["output.hdf"]
+
+        self.rqmt = {
+            "gpu": 1 if device == "gpu" else 0,
+            "cpu": cpu_rqmt,
+            "mem": mem_rqmt,
+            "time": time_rqmt,
+        }
+
+    def tasks(self):
+        yield Task("create_files", mini_task=True)
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    def create_files(self):
+        config = self.create_returnn_config(
+            model_checkpoint=self._model_checkpoint,
+            returnn_config=self._returnn_config,
+            eval_mode=self._eval_mode,
+            log_verbosity=self._log_verbosity,
+            device=self._device,
+        )
+        config.write(self.out_returnn_config_file.get_path())
+
+        cmd = [
+            tk.uncached_path(self.returnn_python_exe),
+            os.path.join(tk.uncached_path(self.returnn_root), "rnn.py"),
+            self.out_returnn_config_file.get_path(),
+        ]
+        util.create_executable("rnn.sh", cmd)
+
+        # check here if model actually exists
+        assert os.path.exists(
+            self._model_checkpoint.index_path.get_path()
+        ), "Provided model does not exists: %s" % str(self._model_checkpoint)
+
+    def run(self):
+        # run everything in a TempDir as writing HDFs can cause heavy load
+        with tempfile.TemporaryDirectory(prefix="work_") as d:
+            print("using temp-dir: %s" % d)
+            call = [
+                tk.uncached_path(self.returnn_python_exe),
+                os.path.join(tk.uncached_path(self.returnn_root), "rnn.py"),
+                self.out_returnn_config_file.get_path(),
+            ]
+
+            # stash a possible exception until we finished copying files from the temp work to the actual
+            # work folder to be able to examine the files
+            error = None
+            try:
+                sp.check_call(call, cwd=d)
+            except Exception as e:
+                print("Run crashed - copy temporary work folder")
+                error = e
+
+            # move log and tensorboard
+            shutil.move(os.path.join(d, "returnn.log"), "returnn.log")
+            tensorboard_dirs = glob.glob(os.path.join(d, "eval-*"))
+            for dir in tensorboard_dirs:
+                shutil.move(dir, os.path.basename(dir))
+
+            # move hdf outputs to output folder
+            for k, v in self.out_hdf_files.items():
+                try:
+                    shutil.move(os.path.join(d, k), v.get_path())
+                except Exception as e:
```

Comment:
`Exception` might be too broad, if you are maybe expecting a `FileNotFoundException`, then only drop those.

Are we even interested in these (partial) outputs if the job did crash? maybe just let it die..